### DISCUSSION TITLE
Impossible for end to end test to detect loading state from DOM elements

### DISCUSSION THREAD

**Discussion by karlhorky** [vercel/next.js#71447] [Help]
### Link to the code that reproduces this issue

https://codesandbox.io/p/devbox/shy-http-8h9hh9?file=%2Fapp%2Fpage.tsx&workspaceId=7ecd71f9-9fcd-424c-8c4e-dfa4601f87d5

### To Reproduce

1. Open the reproduction
2. Click on "2"
3. Try to determine (only from the DOM / browser APIs available in a Playwright end to end test) that the "loading..." on the screen is coming from a loading state

I have been unable to complete step 3 - it appears that it's not possible to detect the loading state without knowing something about the content on the page.

https://github.com/user-attachments/assets/540029a2-5260-4613-821e-0ac97b32480f

![Screenshot 2024-10-17 at 11 23 28](https://github.com/user-attachments/assets/7b862b4a-da0d-48da-bf45-3745dc15e4de)

### Current vs. Expected behavior

Current: impossible to detect loading state from DOM or browser APIs

Expected: possible to detect loading state from DOM or browser APIs

### Provide environment information

```bash
Operating System:
  Platform: linux
  Arch: x64
  Version: #1 SMP PREEMPT_DYNAMIC Sun Aug  6 20:05:33 UTC 2023
  Available memory (MB): 8198
  Available CPU cores: 4
Binaries:
  Node: 20.9.0
  npm: 9.8.1
  Yarn: 1.22.19
  pnpm: 8.10.2
Relevant Packages:
  next: 15.0.0-canary.196 // Latest available version is detected (15.0.0-canary.196).
  eslint-config-next: N/A
  react: 19.0.0-rc-cd22717c-20241013
  react-dom: 19.0.0-rc-cd22717c-20241013
  typescript: 5.3.3
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

Navigation

### Which stage(s) are affected? (Select all that apply)

next dev (local), next start (local)

### Additional context

I'm using Playwright for writing end to end tests

**Comment by samcx** [comment]
Not sure if this is a bug—I'm pretty sure we're testing for loading states via e2e tests.

Can you add playwright testing to your reproduction?

Also, we do have this experimental command, it may help → https://github.com/vercel/next.js/pull/64352.

**Reply by karlhorky** [reply]
Sorry, I forgot to include the screenshot and video that I took - added it to the discussion above now.

> Can you add playwright testing to your reproduction?

Definitely can do that, sure!

However, I don't know what I would add as a test. In the screenshot and video, it shows the DOM, where I can't identify something specific about the `loading.tsx` / loading state.

> I'm pretty sure we're testing for loading states via e2e tests.

Oh that's great, nice!

These tests also include client-side navigations, I guess? (that's what my reproduction above shows)

I'll try to look around in the tests to see if I can find anything.

> Also, we do have this experimental command, it may help → #64352.

Thanks for the additional tip! I'll see whether this also helps.

**Reply by karlhorky** [reply]
> I'm pretty sure we're testing for loading states via e2e tests.

Just looking through the tests in [`test/e2e`](https://github.com/vercel/next.js/tree/canary/test/e2e), I saw this test by @ztanner (from PR https://github.com/vercel/next.js/pull/63768 ):

https://github.com/vercel/next.js/blob/e8885417c75d9cd9cae95f0f008043111696befa/test/e2e/app-dir/dynamic-interception-route-revalidate/dynamic-interception-route-revalidate.test.ts#L23-L24

Another example of the `#loading` id in a different test over here:

https://github.com/vercel/next.js/blob/e8885417c75d9cd9cae95f0f008043111696befa/test/e2e/app-dir/app-prefetch-false-loading/app-prefetch-false-loading.test.ts#L12

But (if I'm understanding correctly) this seems to be testing for a DOM element where the DOM is already known / controlled by the user writing the test (the `#loading` id)

I'm trying to write a test for student apps where I don't control what they put in their loading state (and I also do not control what they put on the following page, so I cannot test for any specific content on that page that will NOT be in their loading state)

I want a Playwright / Puppeteer assertion to be something like this:

```ts
// DO NOT USE, NON-WORKING DEMO CODE

// Wait for the `html` element to have the `__next_loading` attribute
// (loading state is displayed)
await page.waitForSelector('html[__next_loading]');

// Wait for the `html` element to not have `__next_loading` attribute anymore
// (loading state is not displayed anymore)
await page.waitForSelector('html:not([__next_loading])');
```

Maybe there are other tests elsewhere in `test/e2e` that do this a different way, maybe with other Next.js testing tools.

**Reply by karlhorky** [reply]
Actually, I would argue that this may even be considered a bug in Next.js (or maybe it's a bug in the model of single page applications generally) where I cannot click on a link in Puppeteer and wait for the browser to report that it's done navigating, at which point everything is done and rendered and available for the next page.

```ts
const twoLink = await page.$('a[href="/two"]');

// Click on button and wait for navigation (until network idle)
await Promise.all([
  page.waitForNavigation({ waitUntil: 'networkidle0' }),
  twoLink.click(),
]);

// Click on an element that is only on page `/two`
// 💥 Fails with  `Error: No element found for selector a[href="#anchor-only-on-page-two"]`
await page.click('a[href="#anchor-only-on-page-two"]')
```

In this case Puppeteer will fail on not immediately finding the correct element. Can be worked around with waiting for the element (modern testing frameworks like Playwright do auto-waiting and retrying on elements), but it makes code more complicated for automation libraries such as Puppeteer, where auto-waiting and retrying is not built in.

This is (in my understanding) because SPAs will not report the end of the navigation in a way that is visible to the browser automation library. (also not visible to the user - no spinner on the browser tab, etc)

Maybe this needs a new browser API / web standards proposal, if there isn't something already available for this.
